lib/logstash/outputs/elasticsearch.rb

```diff
@@ -300,6 +300,10 @@ def register
     # to build_client down to the Pool class.
     @client = build_client(LicenseChecker.new(@logger))
 
+    # Avoids race conditions in the @data_stream_config initialization (invoking check_data_stream_config! twice).
+    # It's being concurrently invoked by this register method and by the finish_register on the @after_successful_connection_thread
+    data_stream_enabled = data_stream_config?
```

Comment:
A note on `#data_stream_config?` report that:
```
# @note assumes to be running AFTER {after_successful_connection} completed, due ES version checks
```

https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/v11.12.1/lib/logstash/outputs/elasticsearch/data_stream_support.rb#L31-L34

So here can't be invoked

Reply:
Thanks for reviewing @andsel, Not sure If I missed something, but It seems it actually can (and is currently randomly doing it). that's what I tried to explain in my comment here: https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1109#discussion_r1068362267

Reply:
Missed that sorry :-(
A couple of thoughts, the code comment seems to refer to the ES version retrieved during HTTP connection, but as I can see in the code that stays nested in the method `#check_data_stream_config!` (and method invoked inside this method), doesn't seem that it ever checks the version retrieved from the remote ES, to decide if enable or disable the datastream feature. It just checks Logstash version against some static constants.

What' effectively checks the live ES version, is in the `#finish_register` method the `assert_es_version_supports_data_streams`

```ruby
def finish_register
  assert_es_version_supports_data_streams if data_stream_config?
  ...
end  
```

So from my perspective we can remove the [code comment](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/25aedb6e448997c6831839b97221959028917583/lib/logstash/outputs/elasticsearch/data_stream_support.rb#L31)  and keep your fix to avoid race on the `@data_stream_config` field between the "open connector" thread and the one that's executing the `#register`

Reply:
Exactly! I actually already moved the comment to the [assert_es_version_supports_data_streams](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1109/files#diff-56fc556089a712f2ab141b29aa8363f8aaf2f3ee65360c538cc42245a0647ca9R144) method.